lib/btree.js

```diff
@@ -0,0 +1,129 @@
+'use strict';
+
+const DEFAULT_DEGREE = 6; // min degree of b-tree.
+// All vertices except the root have [degree ... 2 * degree] child nodes
+// And [degree - 1 ... 2 * degree - 1] + 1 "empty" elements
+
+class Element {
+  constructor(value, data, child = null) {
+    this.value = value;
+    this.data = data;
+    this.child = child;
+  }
+}
+
+const empty = child => new Element(undefined, null, child);
+
+const splitNode = (parent, index) => {
+  const node = parent[index].child;
+  const len = node.length;
+  const newLeftNode = node.splice(0, len / 2 - 1); // First half
+  const mid = node.shift();
+  newLeftNode.push(empty(mid.child));
+  mid.child = newLeftNode;
+  parent.splice(index, 0, mid);
+  return parent;
+};
+
+const isLeaf = node => !node[0].child;
+
+const binarySearch = (node, value) => {
+  let start = 0;
+  let end = node.length - 1;
+  while (start <= end) {
+    const i = (start + end) >> 1;
+    const item = node[i];
+    const itemValue = item.value;
+    if (value > itemValue) start = i + 1;
+    else if (!(value >= itemValue)) end = i - 1;
+    // We use reverse compare, because the last element have value === undefined
+    else return [true, i];
+  }
+  return [false, start];
+};
+
+function* inorderTraversal(start, finish, currNode) {
+  let [found, startIndex] = binarySearch(currNode, start);
+  if (found) startIndex++;
+  let finishIndex = finish === undefined ?
+    currNode.length - 1 :
+    binarySearch(currNode, finish)[1];
+  if (isLeaf(currNode)) {
+    for (let currIndex = startIndex; currIndex < finishIndex; currIndex++) {
+      yield currNode[currIndex].data;
+    }
+  } else {
+    for (let currIndex = startIndex; currIndex < finishIndex; currIndex++) {
+      const currElement = currNode[currIndex];
+      yield* inorderTraversal(start, finish, currElement.child);
+      yield currElement.data;
+    }
+    yield* inorderTraversal(start, finish, currNode[finishIndex].child);
+  }
+}
+
+class BTree {
+
+  constructor(degree = DEFAULT_DEGREE) {
+    this.root = [empty()];
+    this.minDegree = degree;
+  }
+
+  get(value) {
+    let currNode = this.root;
+    while (currNode) {
+      const [found, i] = binarySearch(currNode, value);
+      if (found) {
+        return currNode[i].data;
+      }
+      currNode = currNode[i].child;
+    }
+    return null;
+  }
+
+  insert(value, data) {
```

Comment:
Could you also add a `replace` boolean flag parameter to just replace the value if it already exists in the tree?

Reply:
I also thought about the flag ``include`` in the function  ``iterator`` which shows: the iterator includes or not the edges of the segment in the result. It makes sense?

Reply:
It seems to me most likely it will be necessary to include either both borders, or not include a single one. So maybe we can just leave one flag that will inform about both borders

Reply:
Well, I'm not so sure, I think we can discuss this later. In the end, it's easy to add such a flag.

Reply:
OK